added splrep, splprep and related methods from scipy

bshall opened the pull request:
I needed to use parametric splines so I thought I might as well make a pull request. There'll be a bit of redundancy with existing code so it may be a good idea to do some refactoring.

kbarbary commented:
Thanks! As you probably realize, these functions mimic the older scipy "tck"-based interface to libdierckx, whereas the existing functionality in Dierckx.jl mimics the newer "object-oriented" interface. We want to stick with mimicing the newer-style interface.

There is some overlap. For example, I think `splev` and `splint` and `splder` are basically duplicates of the existing functions `evaluate(::Spline1D, x)`, `integrate(::Spline1D, a, b)` and `derivative(::Spline1D, x)`.

I think maybe the functions `splrep` and `splprep` could be integrated into the constructor for `Spline1D` (which is basically a container for the `t,c,k` output). They do indeed contain some stuff that isn't yet present in this package, such as periodic boundary conditions. They call a couple different underlying Fortran functions: `parcur_`, `clocur_` and `percur_`; I haven't looked too deeply at the differences between these and `curfit_`. Do you think it would make sense to integrate this into the `Spline1D` constructor? An alternative would be a function with a different name that returns a `Spline1D` instance.

bshall commented:
No problem. Yeah, the "object-oriented" interface is much nicer. I've incorporated `splrep` and `splprep` into the `Spline1D` constructors. I also added a `periodic` keywork. I haven't got around to getting the `roots` function to work with parametric splines yet though.

bshall commented:
On a side note, merging this will be going in a different direction to scipy's BSpline class. The 0.19.0 release notes have this to say
> For multidimensional splines, `c.ndim > 1`, `BSpline` objects are consistent with piecewise polynomials, `scipy.interpolate.PPoly`. This means that `BSpline` objects are not immediately consistent with `scipy.interpolate.splprep`, and one cannot do `>>> BSpline(*splprep([x, y])[0])`. Consult the `scipy.interpolate` test suite for examples of the precise equivalence.

kbarbary commented:
Apologies. I didn't look carefully enough at what `splprep` actually does!

Now that I understand things a bit better, I think we need a different type other than `Spline1D` to represent the output from `parcur_` and `clocur_`: `PSpline1D` or `ParametricSpline`? What do you think?

bshall commented:
We can make a new type if you think that would be a better interface. What I've done is to have 2 constructors, if `Spline1D(x::AbstractVector, y::AbstractVector)` is called then a normal 1D spline is made but if either `Spline1D(x::AbstractMatrix)` or `Spline1D(u::AbstractVector, x::AbstractMatrix)` are called then a parametric spline is constructed.

So it's possible to cram everything into a single type but maybe the interface is a bit clumsy?

kbarbary commented:
The `VecOrMat{Float64}` makes things type-unstable. This could be avoided by making `Spline1D` a parametric type, but I think it makes more sense to just have two separate types. `Spline1D` seems a bit misleading for a spline through N-d space anyway.

bshall commented:
Right, that makes sense. I've added `ParametricSpline` as a new type now. I just need to add the `roots` function now.

kbarbary commented:
Looks a lot better with `ParametricSpline`! Can you remove the `spl*` functions now?

Style-wise, looks mostly good, but prefer 4 space indents on `if...else` blocks, as with other blocks.

bshall commented:
Great. I've removed the spl* functions and fixed the spacing. Let me know if there's anything else.

kbarbary commented:
:+1: Thanks for the revisions.
